From e160da43572c1ac513e0acd4801947e2fb50f4ec Mon Sep 17 00:00:00 2001
From: Xilin Wu <sophon@radxa.com>
Date: Thu, 7 May 2026 22:04:31 +0800
Subject: [PATCH 127/268] net: stmmac: tc956x: Add shutdown callback to suspend
 the MAC

The board sometimes fails to shutdown due to PCIe NoC timeout. It turns
out that the MAC driver sometimes tries to do a MDIO MMIO read when the
PCIe link / bus is already down.

Add shutdown callback to suspend the MAC so the board can shutdown
gracefully.

Signed-off-by: Xilin Wu <sophon@radxa.com>
---
 drivers/net/ethernet/stmicro/stmmac/dwmac-tc956x.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-tc956x.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-tc956x.c
@@ -767,6 +767,17 @@ static void tc956x_dwmac_remove(struct a
 	tc956x_mac_disable(td);
 }
 
+static void tc956x_dwmac_shutdown(struct auxiliary_device *adev)
+{
+	struct device *dev = &adev->dev;
+	int ret;
+
+	ret = stmmac_suspend(dev);
+	if (ret)
+		dev_warn(dev, "failed to suspend MAC during shutdown: %d\n",
+			 ret);
+}
+
 static const struct auxiliary_device_id tc956x_dwmac_ids[] = {
 	{ .name = TC956X_PCIE_DRIVER_NAME "." TC956X_XGMAC_DEV_NAME, },
 	{ },
@@ -777,6 +788,7 @@ static struct auxiliary_driver tc956x_dw
 	.name		= DRIVER_NAME,
 	.probe		= tc956x_dwmac_probe,
 	.remove		= tc956x_dwmac_remove,
+	.shutdown	= tc956x_dwmac_shutdown,
 	.id_table	= tc956x_dwmac_ids,
 	.driver = {
 		.name	= DRIVER_NAME,
